### PR TITLE
Support for Liebert GXE 1-3KVA series

### PR DESCRIPTION
**Pull request by goodspeed34**
Product Breif: https://www.vertiv.com/4a3d25/globalassets/products/critical-power/uninterruptible-power-supplies-ups/vertiv-liebert-gxe-1000-3000va-racktower-230v-single-phase-ups/vertiv-liebert-gxe-datasheet-en-emea---mka4l0ukvlg.pdf

This UPS has a RS232 port and a USB port. The USB port has an ACM interface which functions as a serial port. Both ports can be used managing the device.

## General points

- [ ] Described the changes in the PR submission or a separate issue, e.g.
  known published or discovered protocols, applicable hardware (expected
  compatible and actually tested/developed against), limitations, etc.

- [ ] There may be multiple commits in the PR, aligned and commented with
  a functional change. Notably, coding style changes better belong in a
  separate PR, but certainly in a dedicated commit to simplify reviews
  of "real" changes in the other commits. Similarly for typo fixes in
  comments or text documents.

- [ ] Please star NUT on GitHub, this helps with sponsorships! ;)

## Frequent "underwater rocks" for driver addition/update PRs

- [ ] Revised existing driver families and added a sub-driver if applicable
  (`nutdrv_qx`, `usbhid-ups`...) or added a brand new driver in the other
  case.

- [ ] Did not extend obsoleted drivers with new hardware support features
  (notably `blazer` and other single-device family drivers for Qx protocols,
  except the new `nutdrv_qx` which should cover them all).

- [ ] For updated existing device drivers, bumped the `DRIVER_VERSION` macro
  or its equivalent.

- [ ] For USB devices (HID or not), revised that the driver uses unique
  VID/PID combinations, or raised discussions when this is not the case
  (several vendors do use same interface chips for unrelated protocols).

- [ ] For new USB devices, built and committed the changes for the
  `scripts/upower/95-upower-hid.hwdb` file

- [ ] Proposed NUT data mapping is aligned with existing `docs/nut-names.txt`
  file. If the device exposes useful data points not listed in the file, the
  `experimental.*` namespace can be used as documented there, and discussion
  should be raised on the NUT Developers mailing list to standardize the new
  concept.

- [ ] Updated `data/driver.list.in` if applicable (new tested device info)

## Frequent "underwater rocks" for general C code PRs

- [ ] Did not "blindly assume" default integer type sizes and value ranges,
  structure layout and alignment in memory, endianness (layout of bytes and
  bits in memory for multi-byte numeric types), or use of generic `int` where
  language or libraries dictate the use of `size_t` (or `ssize_t` sometimes).

- [ ] Progress and errors are handled with `upsdebugx()`, `upslogx()`,
  `fatalx()` and related methods, not with direct `printf()` or `exit()`.
  Similarly, NUT helpers are used for error-checked memory allocation and
  string operations (except where customized error handling is needed,
  such as unlocking device ports, etc.)

- [ ] Coding style (including whitespace for indentations) follows precedent
  in the code of the file, and examples/guide in `docs/developers.txt` file.

- [ ] For newly added files, the `Makefile.am` recipes were updated and the
  `make distcheck` target passes.

## General documentation updates

- [ ] Updated `docs/acknowledgements.txt` (for vendor-backed device support)

- [ ] Added or updated manual page information in `docs/man/*.txt` files
  and corresponding recipe lists in `docs/man/Makefile.am` for new pages

- [ ] Passed `make spellcheck`, updated spell-checking dictionary in the
  `docs/nut.dict` file if needed (did not remove any words -- the `make`
  rule printout in case of changes suggests how to maintain it).

## Additional work may be needed after posting this PR

- [ ] Propose a PR for NUT DDL with detailed device data dumps from tests
  against real hardware (the more models, the better).

- [ ] Address NUT CI farm build failures for the PR: testing on numerous
  platforms and toolkits can expose issues not seen on just one system.

- [ ] Revise suggestions from LGTM.COM analysis about "new issues" with
  the changed codebase.

### CODE REVIEW

File: drivers/gxe.c [hunk 1]
```diff
@@ -0,0 +1,426 @@
...
+	case 0xe0:
+		dstate_setinfo("battery.charger.status", "resting");
+		break;
+	case 0xe1:
+	case 0xe2:
+		dstate_setinfo("battery.charger.status", "charging");
+		break;
+	case 0xe3:
+		dstate_setinfo("battery.charger.status", "discharging");
+	}
```
Comment (jimklimov) [inline drivers/gxe.c:261]: I think some compilers' static analyzers (e.g. `clang-16` or newer) would complain about lacks of `break` here, and of `default:` label in general. Also in other `switch/case` blocks here...

**Review by jimklimov — CHANGES_REQUESTED** [review]
Thanks, a very solid first entry, paint me impressed :)

I've posted some review comments, perhaps CI build agents' warnings and errors would add more.

Generally, a `NEWS.adoc` entry would be welcome, and a man page required. Here I would welcome the man page to also have a description or examples how to use both USB and serial connections, given that the driver does not have USB-specific code (does the UPS present itself as just a `/dev/ttyUSB0` sort of serial port in the OS?)

Maybe mention which USB vendor/product IDs can be seen on the device, are they specific to the UPS model, or generic for USB-Serial chips (FTDI etc.)?

**Comment by jimklimov** [comment]
Also, to set up the build with fatal warnings among other things, you can run `./ci_build.sh` and then iterate with `make` to gobble up error fixes as you edit the sources until none are left.

File: drivers/Makefile.am [hunk 2]
```diff
@@ -57,7 +57,7 @@ SERIAL_DRIVERLIST = al175 bcmxcp belkin belkinunv bestfcom	\
  gamatronic genericups isbmex liebert liebert-esp2 masterguard metasys	\
  mge-utalk microdowell microsol-apc mge-shut oneac optiups powercom rhino 	\
  safenet nutdrv_siemens-sitop solis tripplite tripplitesu upscode2 victronups powerpanel \
- blazer_ser ivtscd apcsmart apcsmart-old riello_ser sms_ser bicker_ser
+ blazer_ser ivtscd apcsmart apcsmart-old riello_ser sms_ser bicker_ser gxe
```
[the remark this answers is not in the document]
Comment (goodspeed34) [reply drivers/Makefile.am:60]: Sure, I'm also thinking this.

File: drivers/gxe.c [hunk 3]
```diff
@@ -0,0 +1,426 @@
...
+		fatalx(EXIT_FAILURE, "gxe: not enough data");
+
+	/* UPS Name, 10 bytes */
+	substr_from_asciihex(databuf, recvbuf+13, 20);
+	dstate_setinfo("ups.mfr", "EpsonNetworkPower");
+	dstate_setinfo("ups.model", "%s", databuf);	
+
+	dstate_setinfo("ups.id", "%s", devaddr);
+	
+	dstate_addcmd("test.battery.start");
```
[the remark this answers is not in the document]
Comment (goodspeed34) [reply drivers/gxe.c:192]: Yes, this is still work in progress

**Comment by goodspeed34** [comment]
@jimklimov Thank you for the swift review. This PR is still work in progress. It looks like for now the endian function is not portable and gcc complains snprintf sometimes. I will try to split the checksum and frame functions into an individual header. It looks like that belongs to a protocol heavily used by Vertiv but without an available standard.

Also the UPS's eco mode can be toggled from serial line. But, I can't find a specific instcmd or var for it in NUT. Can you give a pointer for that?

**Comment by goodspeed34** [comment]
> Thanks, a very solid first entry, paint me impressed :)
> 
> I've posted some review comments, perhaps CI build agents' warnings and errors would add more.
> 
> Generally, a `NEWS.adoc` entry would be welcome, and a man page required. Here I would welcome the man page to also have a description or examples how to use both USB and serial connections, given that the driver does not have USB-specific code (does the UPS present itself as just a `/dev/ttyUSB0` sort of serial port in the OS?)

It's a `/dev/ttyACM` device.

> Maybe mention which USB vendor/product IDs can be seen on the device, are they specific to the UPS model, or generic for USB-Serial chips (FTDI etc.)?

Sure, I will check that.

> Also, is this an "individual" contribution or do you work for that company (who is it now - Liebert, Vertiv, Epson...)? Especially if it is their official/sponsored driver, a note in `acknowledgements.txt` may be worthwhile.

It's individual I think. I asked the company for the driver to RE but they said they don't have one and gave me a protocol manual.

> Finally, an entry or two (serial and USB) in `data/driver.list.in`, including a suitable value for the "support level" column, would be great.

OK, I will do the checklist when ready.

**Comment by jimklimov** [comment]
ECO mode is part of some other discussions slowly progressing in different threads. Currently NUT naming space has had no concept for that, so it is worth introducing in some consistent manner so it becomes a sort of de-facto standard that best suits everyone.

Note the definition of "ECO mode" is also vague and may vary between vendors and models. Seems like a marketing buzzword, and may possibly mean an ability of an UPS to switch between what was known as "on-line" (always dual transformation with a battery in the middle) vs. "line-interactive" (glorified power strip, ~10ms gap in an outage, only then battery-fed load) modes at will, at least on some devices.

**Comment by jimklimov** [comment]
> > Also, is this an "individual" contribution or do you work for that company (who is it now - Liebert, Vertiv, Epson...)? Especially if it is their official/sponsored driver, a note in acknowledgements.txt may be worthwhile.

> It's individual I think. I asked the company for the driver to RE but they said they don't have one and gave me a protocol manual.

Sounds decent. Can you please ask if a copy of the protocol file can be published e.g. at https://networkupstools.org/ups-protocols.html ?

**Comment by goodspeed34** [comment]
> > > Also, is this an "individual" contribution or do you work for that company (who is it now - Liebert, Vertiv, Epson...)? Especially if it is their official/sponsored driver, a note in acknowledgements.txt may be worthwhile.
> 
> > It's individual I think. I asked the company for the driver to RE but they said they don't have one and gave me a protocol manual.
> 
> Sounds decent. Can you please ask if a copy of the protocol file can be published e.g. at https://networkupstools.org/ups-protocols.html ?

Maybe not... There is a confidentiality notice in the message and it's all in Chinese.

**Comment by jimklimov** [comment]
> Maybe not... There is a confidentiality notice in the message and it's all in Chinese.

Well, I am not sure what the social conventions are then... would it *hurt* to *ask* if they have a publishable variant (and perhaps in English)?

**Comment by goodspeed34** [comment]
> > Maybe not... There is a confidentiality notice in the message and it's all in Chinese.
> 
> Well, I am not sure what the social conventions are then... would it _hurt_ to _ask_ if they have a publishable variant (and perhaps in English)?

It will try my best letting that happen

**Comment by goodspeed34** [comment]
@jimklimov As they are standardized, I'm trying to put all the constants and functions into a header https://github.com/networkupstools/nut/pull/2629/commits/b301f7971d74cf8e0c9303f165b014066ad26ada#diff-38aa0ff4c43ee1455b1c9ad39462ebdd4eb883402579bc5d812af64adf42701a. However, when included and not all variables were used, complier complains unused variables. How to solve that?

**Comment by goodspeed34** [comment]
The corresponding protocol is attached. Sadly, I can't find an English one.

[YDT1363.3-2014通信局(站)电源、空调及环境集中监控管理系统第3部分：前端智能设备协议.pdf](https://github.com/user-attachments/files/17008757/YDT1363.3-2014.3.pdf)


> > Maybe not... There is a confidentiality notice in the message and it's all in Chinese.
> 
> Well, I am not sure what the social conventions are then... would it _hurt_ to _ask_ if they have a publishable variant (and perhaps in English)?

**Comment by goodspeed34** [comment]
poll_interval needs revision...

File: drivers/liebert-gxe.c [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+	for (retry = 0; retry < PROBE_RETRIES; retry++) {
+		ret = ydn23_frame_send(upsfd, &sendframe);
+		if (ret <= 0) continue;
+		ret = ydn23_frame_read(upsfd, &recvframe, 1, 0);
+		if (ret > 0)
+			break;
+	}
+
+	if (ret <= 0)
+		fatal_with_errno(EXIT_FAILURE, "gxe: failed reading response");
```
Comment (jimklimov) [inline drivers/liebert-gxe.c:412]: Saw this complaint pattern recently, not sure OTOH in which PR:

````
[1mliebert-gxe.c:411:6: [0m[0;1;31merror: [0m[1mvariable 'ret' may be uninitialized when used here [-Werror,-Wconditional-uninitialized][0m
        if (ret <= 0)
[0;1;32m            ^~~
[0m[1mliebert-gxe.c:398:16: [0m[0;1;30mnote: [0minitialize the variable 'ret' to silence this warning[0m
        int retry, ret;
[0;1;32m                      ^
[0m[0;32m                       = 0
[0m1 error generated.
make[1]: *** [Makefile:2113: liebert-gxe.o] Error 1
````

It seems `clang` static analysis (across a ton of compiler versions, 7 to 18 at least) can not guarantee that we would generally get into the loop and so assign `ret`, although it seems obvious to human eye that we should. So it asks to pre-initialize e.g. `int ret = -1` in line 398 to surely use a deterministic value in line 411.

I'd shrug it off as a compiler bug, but:
* it bites us in CI which we want to pass warnings-free anyway
* it does not hurt much to add the assignment (just annoys to do so :) )
* maybe the compiler does know better about how it optimizes something away...

Comment (jimklimov) [reply drivers/liebert-gxe.c]: Posted the fix commit.

**Comment by goodspeed34** [comment]
@jimklimov Is it possible to have `poll_interval` changed according to user's configuration? As mentioned in the code, serial line sometimes ignores the message if sending too quick therefore stale the driver.

**Comment by goodspeed34** [comment]
It seems CI bugged out from the forced-push...

**Review by jimklimov — APPROVED** [review]
Super! Thanks!

**Comment by jimklimov** [comment]
@goodspeed34 : By the way, one more question: since the beginning of time, there already was a `drivers/liebert.c` with claimed support for "Liebert UPStation GXE" "with MultiLink cable". Out of curiosity - did you check if it worked for your device?

I do see that the old driver is very basic and less functional than yours; just wondering if it can be treated as obsolete in favor of new development, or if they cover different grounds (e.g. cables/protocols) and should both exist?

**Comment by goodspeed34** [comment]
> @goodspeed34 : By the way, one more question: since the beginning of time, there already was a `drivers/liebert.c` with claimed support for "Liebert UPStation GXE" "with MultiLink cable". Out of curiosity - did you check if it worked for your device?
> 
> I do see that the old driver is very basic and less functional than yours; just wondering if it can be treated as obsolete in favor of new development, or if they cover different grounds (e.g. cables/protocols) and should both exist?

It doesn't work and shows always OL state.

**Comment by goodspeed34** [comment]
> > @goodspeed34 : By the way, one more question: since the beginning of time, there already was a `drivers/liebert.c` with claimed support for "Liebert UPStation GXE" "with MultiLink cable". Out of curiosity - did you check if it worked for your device?
> > I do see that the old driver is very basic and less functional than yours; just wondering if it can be treated as obsolete in favor of new development, or if they cover different grounds (e.g. cables/protocols) and should both exist?

It doesn't work and shows always OL state. (This UPS model only recognize the data frame in `ydn23.h`, there won't be any responses if data gets sent in `liebert.c` way. It looks like that driver lacks a probing method and report false positively) I think there are for different models and I can't tell if it's obsolete or not.

**Comment by jimklimov** [comment]
Thanks for clarification, much appreciated!

**Comment by jimklimov** [comment]
Hm, did I miss a data dump as an example of the new driver's work in the comments above, or could you kindly post one? or a PR to nut-ddl repo right away :)

**Comment by goodspeed34** [comment]
I don't have my laptop with me so I'm afraid I can't raise a PR. Here is the dump for Liebert GXE 01k00TS1101C00:
```
goodspeed@radiata:~$ upsc emerson
Init SSL without certificate database
battery.charger.status: charging
battery.runtime.low: 54.00
battery.voltage: 40.85
device.mfr: EmersonNetworkPower
device.model: HE1T0010
device.type: ups
driver.debug: 0
driver.flag.allow_killpower: 0
driver.name: liebert-gxe
driver.parameter.pollinterval: 5
driver.parameter.port: /dev/ttyACM0
driver.parameter.synchronous: auto
driver.state: quiet
driver.version: 2.8.2.1
driver.version.internal: 0.01
input.frequency: 49.99
input.transfer.bypass.high: 253.000000
input.transfer.bypass.low: 120
input.voltage: 223.22
output.current: 0.66
output.frequency: 49.99
output.frequency.nominal: 50
output.voltage: 225.37
output.voltage.nominal: 220
ups.id: 01
ups.mfr: EmersonNetworkPower
ups.model: HE1T0010
ups.power: 140
ups.realpower: 120
ups.status: OL BYPASS
ups.test.interval: 0
ups.test.result: Idle
```

> Hm, did I miss a data dump as an example of the new driver's work in the comments above, or could you kindly post one? or a PR to nut-ddl repo right away :)

**Comment by jimklimov** [comment]
Thanks!

**Pull request merged**